lib/util/src/timer.c

```diff
@@ -197,21 +201,32 @@ hse_timer_init(void)
 void
 hse_timer_fini(void)
 {
-    struct timer_list *t, *next;
-
     if (!timer_wq)
         return;
 
+    /* There shouldn't be any pending timers, but if there
+     * are we'll log a message and force them to expire.
+     */
+    while (1) {
+        struct timer_list *first;
+
+        timer_lock();
+        first = timer_first();
+        if (first) {
+            log_err("timer %p abandoned, func %p, data %lu, expires in %lu jiffies\n",
+                    first, first->function, first->data, first->expires - jiffies);
+            first->expires = jiffies;
+            assert(0);
+        }
+        timer_unlock();
+
+        if (!first)
+            break;
+
+        usleep(USEC_PER_SEC);
```

Comment:
Any reason not to just use sleep(1)?

Reply:
I think I originally slept much shorter, but then it would spam the log if there were pending timers (which is a bug, but noneless not very useful).  Longer term we need to get rid of all sleep/msleep/usleep calls and replace them with our own wrapper which ensures they call nanosleep to guarantee they don't interact with the app in some undersirable way...

Reply:
Might want to create a ticket.

Reply:
Good idea, it's been on my mental list for quite a while.  I have been quietly trying to replace msleep calls when I run across them, but there are still a few...

Reply:
You seem to have a lot of ideas, so always best to get them down, so they can be prioritized effectively.

Reply:
Ha!  Well, six are good, half-a-dozen are bad, so I keep 'em to myself until I can no long stand the irritation :-)